token-lending/program/src/state/reserve.rs

```diff
@@ -385,43 +330,44 @@ pub struct ReserveLiquidity {
     pub mint_decimals: u8,
     /// Reserve liquidity supply address
     pub supply_pubkey: Pubkey,
+    /// Reserve liquidity fee receiver address
+    pub fee_receiver: Pubkey,
+    /// Optional reserve liquidity aggregator state account
+    pub aggregator: COption<Pubkey>,
+    /// Reserve liquidity cumulative borrow rate
+    pub cumulative_borrow_rate_wads: Decimal,
+    /// Reserve liquidity median price in quote currency
+    pub median_price: u64,
     /// Reserve liquidity available
     pub available_amount: u64,
     /// Reserve liquidity borrowed
     pub borrowed_amount_wads: Decimal,
 }
 
 impl ReserveLiquidity {
-    /// New reserve liquidity info
-    pub fn new(mint_pubkey: Pubkey, mint_decimals: u8, supply_pubkey: Pubkey) -> Self {
-        Self {
-            mint_pubkey,
-            mint_decimals,
-            supply_pubkey,
-            available_amount: 0,
-            borrowed_amount_wads: Decimal::zero(),
-        }
-    }
-
     /// Calculate the total reserve supply including active loans
     pub fn total_supply(&self) -> Result<Decimal, ProgramError> {
         Decimal::from(self.available_amount).try_add(self.borrowed_amount_wads)
     }
 
-    /// Add new borrow amount to total borrows
-    pub fn borrow(&mut self, borrow_amount: u64) -> ProgramResult {
-        if borrow_amount > self.available_amount {
+    /// Subtract borrow amount from available liquidity and add to borrows
+    pub fn borrow(&mut self, borrow_amount: Decimal) -> ProgramResult {
+        // @TODO: should this be rounded differently?
+        let borrow_rounded = borrow_amount.try_ceil_u64()?;
+        if borrow_rounded > self.available_amount {
             return Err(LendingError::InsufficientLiquidity.into());
         }
 
-        self.available_amount -= borrow_amount;
-        self.borrowed_amount_wads = self
-            .borrowed_amount_wads
-            .try_add(Decimal::from(borrow_amount))?;
+        self.available_amount = self
+            .available_amount
+            .checked_sub(borrow_rounded)
+            .ok_or(LendingError::MathOverflow)?;
+        self.borrowed_amount_wads = self.borrowed_amount_wads.try_add(borrow_amount)?;
+
         Ok(())
     }
 
```

Comment:
@jstarry we pass in a `Decimal` `borrow_amount` now since it includes calculated fees. How should we round the number to subtract from `available_amount`?